hail/src/main/scala/is/hail/expr/ir/Emit.scala

```diff
@@ -2126,6 +2127,92 @@ private class Emit(
           }
         }
 
+      case x@NDArrayConcat(nds, axis) =>
+        val inputType = coerce[PArray](nds.pType2)
+        val inputNDType = coerce[PNDArray](inputType.elementType)
+
+        val ndType = coerce[PNDArray](x.pType2)
+        val codeNDs = emit(nds, env, er, None)
+
+        val inputArray = mb.newField[Long]
+        val n = mb.newField[Int]
+        val i = mb.newField[Int]
+
+        val loadAndValidateArray = Code(
+          inputArray := codeNDs.value[Long],
+          n := inputType.loadLength(inputArray),
+          (n < 1).orEmpty(Code._fatal("need at least 1 array to concatenate")))
+
+        val (missingSetup: Code[Unit @unchecked], missing: Code[Boolean @unchecked], setupShape: Code[Unit @unchecked]) = (inputType.required, inputNDType.required) match {
+          case (true, true) => (Code._empty, const(false), Code(
+            codeNDs.setup,
+            codeNDs.m.orEmpty(Code._fatal("required NDArray can't be missing")),
+            loadAndValidateArray))
+          case (false, true) => (codeNDs.setup, codeNDs.m, loadAndValidateArray)
+          case _ =>
+            val m = mb.newField[Boolean]
+            val setup = Code(
+              codeNDs.setup,
+              m := codeNDs.m,
+              loadAndValidateArray,
+              i := 0,
+              Code.whileLoop(i < n,
+                m := m | inputType.isElementMissing(inputArray, i),
+                i := i + 1))
+            (setup, m.load(), Code._empty)
+        }
+
+        val localDim = mb.newField[Long]
+        val outputShape = Array.tabulate(ndType.nDims) { idx =>
+          if (idx == axis) {
+            Code(
+              localDim := inputNDType.dimensionLength(inputType.loadElement(inputArray, 0), idx),
+              i := 1,
+              Code.whileLoop(i < n,
+                localDim := localDim + inputNDType.dimensionLength(inputType.loadElement(inputArray, i), idx),
+                i := i + 1),
+              localDim)
+          } else {
+            Code(
+              localDim := inputNDType.dimensionLength(inputType.loadElement(inputArray, 0), idx),
+              i := 1,
+              Code.whileLoop(i < n,
+                inputNDType.dimensionLength(inputType.loadElement(inputArray, i), idx).cne(localDim).orEmpty(Code._fatal("mismatched dims")),
```

Comment:
Mention something here about the fact that this error is coming from the concat operation. Stacktrace will show the error as coming from a codegen method so it won't be obvious where the problem in a complicated NDArray pipeline is.